awx/main/models/unified_jobs.py

```diff
@@ -912,36 +906,105 @@ def result_stdout_text(self, value):
-                    return self.result_stdout_raw_handle(attempt=attempt + 1)
+            # If enforce_max_bytes = False, that means they're downloading
+            # the entire file.  To avoid ballooning memory, let's write the
+            # stdout content to a temporary disk location
+            if not os.path.exists(settings.JOBOUTPUT_ROOT):
+                os.makedirs(settings.JOBOUTPUT_ROOT)
+            fd = tempfile.NamedTemporaryFile(
+                prefix='{}-{}-'.format(self.model_to_str(), self.pk),
+                suffix='.out',
+                dir=settings.JOBOUTPUT_ROOT
+            )
+
+        # Before the addition of event-based stdout, older versions of
+        # awx stored stdout as raw text blobs in a certain database column
+        # (`main_unifiedjob.result_stdout_text`)
+        # For older installs, this data still exists in the database; check for
+        # it and use if it exists
+        legacy_stdout_text = self.result_stdout_text
+        if legacy_stdout_text:
+            if enforce_max_bytes and len(legacy_stdout_text) > max_supported:
+                raise StdoutMaxBytesExceeded(len(legacy_stdout_text), max_supported)
+            fd.write(legacy_stdout_text)
+            if hasattr(fd, 'name'):
+                fd.flush()
+                return open(fd.name, 'r')
+            else:
+                # we just wrote to this StringIO, so rewind it
+                fd.seek(0)
+                return fd
+        else:
+            # Note: the code in this block _intentionally_ does not use the
+            # Django ORM because of the potential size (many MB+) of
+            # `main_jobevent.stdout`; we *do not* want to generate queries
+            # here that construct model objects by fetching large gobs of
+            # data (and potentially ballooning memory usage); instead, we
+            # just want to write concatenated values of a certain column
+            # (`stdout`) directly to a file
+
+            with connection.cursor() as cursor:
+                tablename = self._meta.db_table
+                related_name = {
+                    'main_job': 'job_id',
+                    'main_adhoccommand': 'ad_hoc_command_id',
+                    'main_projectupdate': 'project_update_id',
+                    'main_inventoryupdate': 'inventory_update_id',
+                    'main_systemjob': 'system_job_id',
+                }[tablename]
+
+                if enforce_max_bytes:
+                    # detect the length of all stdout for this UnifiedJob, and
+                    # if it exceeds settings.STDOUT_MAX_BYTES_DISPLAY bytes,
+                    # don't bother actually fetching the data
+                    cursor.execute(
+                        "select sum(length(stdout)) from {} where {}={}".format(
+                            tablename + 'event',
+                            related_name,
+                            self.id
+                        )
+                    )
```

Comment:
Won't this result in a `StdoutMaxBytesExceeded` error if you attempt to retrieve a modestly-sized selection from a large standard out? I would eventually want to get to something like:

```python
total_bytes_lower_limit = JobEvent.objects.filter(
    '{}_id'.format(related_name)=self.id, end_line__gt=start_line, start_line__lt=end_line
).annotate(byte_est=Length('stdout')).aggregate(Sum('byte_est'))
```

If I got that right, this would give a lower-limit estimate of the total bytes of the section being requested. There's still a minor chance (and insignificant here) that a multi-line job event would push the returned content over the limit, because these methods only measure event-by-event. Since we're only trying to avoid OOM type of situations, that seems fine to me. You could change it to an upper-limit estimate by switching the `__gt` and `__lt` filters to `__gte` and `__lte` filters, but we would want to test the lower limit, because this only needs to be approximate.

Reply:
You mean via `start_line` and `end_line`?  It _would_ prevent you from taking a slice of a stdout block that was larger than `settings.STDOUT_MAX_BYTES_DISPLAY`, but that's how it works _today_.

Your code example is using the ORM, so I'd want to be certain that `JobEvent.query.filter` it didn't side load the `stdout` column into memory incidentally (similar to our `UnifiedJob.result_stdout_text` performance problem).

Also, using `start_line__gt` and `end_line__lt` on the query in this way may not end up giving us all of the context we need to continue providing this `start_actual` + `end_actual` interface:

https://github.com/ansible/awx/blob/devel/awx/main/models/unified_jobs.py#L984

As you can see, start + end line is _complicated_ :/.  I'd prefer to improve the `start_line` and `end_line` implementation in a separate PR, especially considering this is becoming a deprecated endpoint.

Reply:
> Your code example is using the ORM, so I'd want to be certain that JobEvent.query.filter didn't load the stdout column into memory incidentally 

No problem - I like to test these by adding the computation to the API root view locally and viewing what happens via the debug toolbar. In this case, I have to use hard-coded ids, but that doesn't lose any analytical value.

in views

```python
        data['bytes'] = total_bytes = JobEvent.objects.filter(
            job_id=41, end_line__gte=4, start_line__lte=9
        ).annotate(byte_est=Length('stdout')).aggregate(Sum('byte_est'))['byte_est__sum']
```

the SQL

```sql
SELECT
   SUM("byte_est") 
FROM
   (
      SELECT
         "main_jobevent"."id" AS Col1,
         LENGTH("main_jobevent"."stdout") AS "byte_est" 
      FROM
         "main_jobevent" 
      WHERE
         (
            "main_jobevent"."end_line" >= 4 
            AND "main_jobevent"."start_line" <= 9 
            AND "main_jobevent"."job_id" = 41
         )
      GROUP BY
         "main_jobevent"."id",
         LENGTH("main_jobevent"."stdout")
   )
   subquery
```

The query plan:

https://explain.depesz.com/s/36UQ

And the toolbar will tell me the actual returned result, which is just "408".

Also, posting this for public consumption:

https://gist.github.com/AlanCoding/1a2719b2cf442048d86c2299ff6f04b0